04-onion-routing.md

```diff
@@ -1052,25 +1055,62 @@ The erring node then generates a new key, using the key type `ammag`.
-return message, because of course, it was the originator of the corresponding
-forward packet.
-When an origin node receives an error message matching a transfer it initiated
-(i.e. it cannot return-forward the error any further) it generates the `ammag`
-and `um` keys for each hop in the route.
-It then iteratively decrypts the error message, using each hop's `ammag`
-key, and computes the HMAC, using each hop's `um` key.
-The origin node can detect the sender of the error message by matching the
-`hmac` field with the computed HMAC.
+1. data:
+    * [`20*u32`:`htlc_hold_times`]
+    * [`210*sha256[..4]`:`truncated_hmacs`]
 
-The association between the forward and return packets is handled outside of
-this onion routing protocol, e.g. via association with an HTLC in a payment
-channel.
+The field `htlc_hold_times` contains the htlc hold time in milliseconds for each hop. The sender can use this information to score nodes on latency. Nodes along the path that lack accurate timing information may simply report a value of zero. In such cases, the sender should distribute any potential latency penalty across multiple nodes. This encourages path nodes to provide timing data to avoid being held responsible for the high latency of other nodes.
+
+The erring node puts its hold time at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum supported number of hops in a route (20).
+
+The field `truncated_hmacs` contains truncated authentication codes series for each hop, with the
+same `um` key that is used for `hmac` in the return packet. Regular 32 byte HMACs are
+truncated to the first 4 bytes to save space.
+
+In theory this truncation makes it possible for malicious nodes to guess the
+HMAC. However, game theory is against them because a wrong guess will get
+them penalized.
+
+The size of the field is based on the maximum number of hops in a route (20) and
+the truncated HMAC size (4 bytes). Each hop adds 20 HMACs, one for each possible
+position that the hop could be at in the path. This is necessary because only
+the sender knows the position of each hop in the path.
+
+It is not required to store all 20 * 20 = 400 HMACs. The node in position 0 needs to store an HMAC for every one of the 20 positions. For the node in position 1, the maximum remaining route length is only 19. For that position just 19 HMACs need to be kept, and so on. This makes for a total number of HMACs of 20+19+18+...+1 = 210 HMACs.
+
+The layout of the `hmacs` field is shown below. The actual format is much longer,
+but for readability the format is described as if the maximum route length would
+be just three hops.
+
+`hmac_0_2` | `hmac_0_1`| `hmac_0_0`| `hmac_1_1`| `hmac_1_0`| `hmac_2_0`
+
+`hmac_x_y` is the hmac added by node `x` (counted from the node that is
+currently handling the failure message) assuming that this node is `y` hops
+away from the erring node.
+
+Each HMAC covers the following data:
```

Comment:
"covers" is a bit ambiguous, make it explicit that these things are concatenated in this order. Same below for hold times and hmacs. For hold times, you can say that we take the first `(y + 1) * 4` bytes of `htlc_hold_times`.